crates/extensions/tedge_config_manager/src/actor.rs

```diff
@@ -297,28 +299,90 @@ impl ConfigManagerActor {
         &mut self,
         topic: &str,
         result: DownloadResult,
-    ) -> Result<(), ChannelError> {
-        if let Some(ConfigOperation::Update(mut request)) = self.pending_operations.remove(topic) {
-            let topic = Topic::new_unchecked(topic);
-            match result {
-                Ok(response) => {
-                    request.successful(response.file_path.as_path().to_str().unwrap());
-                    info!(
-                        "Config Update request processed for config type: {}.",
-                        request.config_type
+    ) -> Result<(), ConfigManagementError> {
+        let Some(ConfigOperation::Update(mut request)) = self.pending_operations.remove(topic)
+        else {
+            return Ok(());
+        };
+
+        let topic = Topic::new_unchecked(topic);
+
+        let response = match result {
+            Ok(response) => response,
+            Err(err) => {
+                let error_message = format!("Handling of operation failed with: {}", err);
+                request.failed(&error_message);
+                error!("{}", error_message);
+                self.publish_command_status(&topic, &ConfigOperation::Update(request))
+                    .await?;
+                return Ok(());
+            }
+        };
+
+        // new config was downloaded into tmpdir, we need to write it into destination using tedge-write
+        let file_entry = self
+            .plugin_config
+            .get_file_entry_from_type(&request.config_type)?;
+
+        // when testing, `tedge-agent` binary is not available. need to do the switch here, because
+        // conditional compilation in tedge-write crate here doesn't work
+        #[cfg(test)]
+        {
+            std::fs::copy(&response.file_path, &file_entry.path)?;
+        }
+        #[cfg(not(test))]
+        {
+            let mode = file_entry.file_permissions.mode;
+            let user = file_entry.file_permissions.user.as_deref();
+            let group = file_entry.file_permissions.group.as_deref();
```

Comment:
One very annoying thing: in unit tests, `tedge-write` binary is not present, so instead of spawning a new process, we can just copy the file as usual, but I'd like to do it in `tedge-write` instead of here.

I've tried using `#[cfg(test)]` for changing `copy` function to not spawn a process, but when unit testing, only the current crate is running on a `test` target, so it wouldn't work. Now I'm trying to set up a `testing` feature that only gets used inside tests, but then when doing `cargo check --all-features` it generates warning because the actual implementation is unused.

What would be the best way to provide a testing implementation that doesn't spawn a process?

Reply:
I would try to reduce the extent of this conditional. What is tested is too different of what is actually used.

Could the `CopyOptions` be built and used in both cases with a single difference in the constructor, a flag to either spawn a process or invoke a method (e.g. `tedge_write::bin::run()`).

Reply:
You pushed a fix along these lines while I was writing this comment. Slightly different, but working fine.